### PR TITLE
Editor gizmo upgrade

### PR DESCRIPTION
**Pull request by Jomosoto**
Implements parts of #166 

# Summary

- Change the model to a conventional style
- Make gizmo fixed size (independant of distance to the camera)
- Make the movement axes draggable
- Add rotation for all axes
- Fix a bug, where you can't interact with the gizmo when it's inside of objects

https://github.com/Libre-TrainSim/Libre-TrainSim/assets/40580158/018f51f4-0852-492e-9cd2-c4a0365c1299

---

# Movement

The difference to the start position is calculated by using the law of sines on the triangle between the origin of the camera, the position where the gizmo is grabbed:

grafik

This is the best approach I found, however this is pretty long in code and I feel like there must be something shorter. Please tell me if you know a better approach.

# Rotation

For the rotation, the player circles with the mouse pointer around the object, like in the the godot editor.

The angle to rotate is calculated like this:

grafik

--- 

To fix the bug, where you couldn't interact with the gizmo when it's inside of objects I moved the interaction to physic layer 2.

### CODE REVIEW

**Comment by HaSa1002** [comment]
Nice. I am gonna check out your work hopefully tomorrow. I haven't yet looked into your code in depth but it didn't seem overly verbose. But you couldn't project the cursor into the world? Anyway, I am too tired this evening

File: src/Editor/Modules/Gizmo.tscn [hunk 1]
```diff
@@ -47,81 +46,130 @@ emission_energy = 1.0
...
...455, 1.8258, 18.5375, -1.5455, 1.8258, 16.4873, -0.9552, 3.2795, 18.2692, -1.5455, 3.634, 18.5375, -1.5455, 1.8258, 18.2692, -1.5455, 3.634, 19.6549, 0, 1.9358, 19.6549, 0, 1.9358, 18.2692, -1.5455, 3.634, 19.3705, 0, 3.853, 19.75, 0, 0, 19.6549, 0, 1.9358, 18.6272, 1.5455, 0, 18.6272, 1.5455, 0, 19.6549, 0, 1.9358, 18.5375, 1.5455, 1.8258, 18.6272, 1.5455, 0, 18.5375, 1.5455, 1.8258, 16.8103, 0.9552, 0, 16.8103, 0.9552, 0, 18.5375, 1.5455, 1.8258, 16.7294, 0.9552, 1.6477, 16.8103, 0.9552, 0, 16.7294, 0.9552, 1.6477, 16.8103, -0.9552, 0, 16.8103, -0.9552, 0, 16.7294, 0.9552, 1.6477, 16.7294, -0.9552, 1.6477, 16.8103, -0.9552, 0, 16.7294, -0.9552, 1.6477, 18.6272, -1.5455, 0, 18.6272, -1.5455, 0, 16.7294, -0.9552, 1.6477, 18.5375, -1.5455, 1.8258, 18.6272, -1.5455, 0, 18.5375, -1.5455, 1.8258, 19.75, 0, 0, 19.75, 0, 0, 18.5375, -1.5455, 1.8258, 19.6549, 0, 1.9358, 0, 1.5455, 18.6272, 0, 0, 19.75, 0, 0.9552, 16.8103, 0, 0.9552, 16.8103, 0, 0, 19.75, 0, -0.9552, 16.8103, 0, -0.9552, 16.8103, 0, 0, 19.75, 0, -1.5455, 18.6272, 18.6272, -1.5455, 0, 19.75, 0, 0, 16.8103, -0.9552, 0, 16.8103, -0.9552, 0, 19.75, 0, 0, 16.8103, 0.9552, 0, 16.8103, 0.9552, 0, 19.75, 0, 0, 18.6272, 1.5455, 0 )
```
Comment (HaSa1002) [inline src/Editor/Modules/Gizmo.tscn]: Oh nice. Serialised vector data xD

File: src/Editor/Scripts/EditorHUD.gd [hunk 2]
```diff
@@ -96,12 +98,24 @@ func _on_z_value_changed(value):
...
 	if is_instance_valid(selected_object):
 		selected_object.rotation.y = deg2rad(value)
 
 
+func _on_z_rot_value_changed(value):
+	var selected_object = get_parent().selected_object
+	if is_instance_valid(selected_object):
+		selected_object.rotation.z = deg2rad(value)
+
+
```
Comment (HaSa1002) [inline src/Editor/Scripts/EditorHUD.gd]: Please type these values as float

File: src/Editor/Scripts/Gizmo.gd [hunk 3]
```diff
@@ -5,96 +5,164 @@ var x_active := false
...
-
-	if event is InputEventMouseButton and event.button_index == BUTTON_LEFT:
+var y_rot_hovered := false
+var z_rot_hovered := false
+
+var start_position: Vector3
+var grab_position: Vector3
+
+func _unhandled_input(event: InputEvent) -> void:
+	if event is InputEventMouseMotion and (x_active or y_active or z_active):
```
Comment (HaSa1002) [inline src/Editor/Scripts/Gizmo.gd]:
```suggestion
	var mm := event as InputEventMouseMotion
	if mm != null and (x_active or y_active or z_active):
```

File: src/Editor/Scripts/Gizmo.gd [hunk 4]
```diff
@@ -5,96 +5,164 @@ var x_active := false
...
+		
+		# The position relative to the start position is calculated by using the law of sines
+		# on the triangle between the origin of the camera, the position where the gizmo is grabbed
+		# and the position where the grabbed point is moved to
+		
+		var direction = Vector3(1 if x_active else 0, 1 if y_active else 0, 1 if z_active else 0)
```
Comment (HaSa1002) [inline src/Editor/Scripts/Gizmo.gd]:
```suggestion
		var direction := Vector3(x_active, y_active, z_active)
```

Conversion bool to float can be implicit

File: src/Editor/Scripts/Gizmo.gd [hunk 5]
```diff
@@ -5,96 +5,164 @@ var x_active := false
...
+		
+		var camera = get_viewport().get_camera()
+		
+		var grab_position_on_screen = camera.unproject_position(grab_position)
+		var direction_on_screen = (camera.unproject_position(grab_position + direction) - grab_position_on_screen).normalized()
```
Comment (HaSa1002) [inline src/Editor/Scripts/Gizmo.gd]:
```suggestion
		var camera := get_viewport().get_camera()
		
		var grab_position_on_screen := camera.unproject_position(grab_position)
		var direction_on_screen := (camera.unproject_position(grab_position + direction) - grab_position_on_screen).normalized()
```

File: src/Editor/Scripts/Gizmo.gd [hunk 6]
```diff
@@ -5,96 +5,164 @@ var x_active := false
...
+		
+		# If the axis is dragged in the negative direction
+		if (event.position - grab_position_on_screen).dot(direction_on_screen) < 0:
```
Comment (HaSa1002) [inline src/Editor/Scripts/Gizmo.gd]:
```suggestion
		if (mm.position - grab_position_on_screen).dot(direction_on_screen) < 0:
```

File: src/Editor/Scripts/Gizmo.gd [hunk 7]
```diff
@@ -5,96 +5,164 @@ var x_active := false
...
+			direction *= -1
+		
+		# The object has to be moved to the perpendicular of the mouse position on the axis on screen
+		# We calculate this point:
+		var perpendicular_screen_point = (event.position - grab_position_on_screen).dot(direction_on_screen) * direction_on_screen + grab_position_on_screen;
+		
+		var cam_to_grab_position = grab_position - camera.global_transform.origin
+		var cam_to_new_position = camera.project_ray_normal(perpendicular_screen_point)
```
Comment (HaSa1002) [inline src/Editor/Scripts/Gizmo.gd]:
```suggestion
		var perpendicular_screen_point := (mm.position - grab_position_on_screen).dot(direction_on_screen) * direction_on_screen + grab_position_on_screen;
		
		var cam_to_grab_position := grab_position - camera.global_transform.origin
		var cam_to_new_position := camera.project_ray_normal(perpendicular_screen_point)
```

File: src/Editor/Scripts/Gizmo.gd [hunk 8]
```diff
@@ -5,96 +5,164 @@ var x_active := false
...
+		
+		# The angle at the camera between the vectors to the grab position and the new position
+		var angle_grab_cam_new = cam_to_grab_position.angle_to(cam_to_new_position)
+		# The angle at the object between the vectors to the camera and the new position
+		var angle_cam_object_new = PI - cam_to_grab_position.angle_to(direction)
```
Comment (HaSa1002) [inline src/Editor/Scripts/Gizmo.gd]:
```suggestion
		var angle_grab_cam_new := cam_to_grab_position.angle_to(cam_to_new_position)
		# The angle at the object between the vectors to the camera and the new position
		var angle_cam_object_new := PI - cam_to_grab_position.angle_to(direction)
```

File: src/Editor/Scripts/Gizmo.gd [hunk 9]
```diff
@@ -5,96 +5,164 @@ var x_active := false
...
+		
+		# Limit if the moused is moved beyond the limits of the axis
+		if angle_grab_cam_new + angle_cam_object_new >= PI:
+			angle_grab_cam_new = PI - angle_cam_object_new - 0.01
+		
+		# We get the diff with the law of sines
+		var diff = cam_to_grab_position.length() \
```
Comment (HaSa1002) [inline src/Editor/Scripts/Gizmo.gd]:
```suggestion
		var diff := cam_to_grab_position.length() \
```

File: src/Editor/Scripts/Gizmo.gd [hunk 10]
```diff
@@ -5,96 +5,164 @@ var x_active := false
...
+			/ sin(PI - angle_grab_cam_new - angle_cam_object_new) \
+			* sin(angle_grab_cam_new)
+		
+		get_parent().translation = start_position + diff * direction
+		
+	elif event is InputEventMouseMotion and (x_rot_active or y_rot_active or z_rot_active):
```
Comment (HaSa1002) [inline src/Editor/Scripts/Gizmo.gd]:
```suggestion
	elif mm != null and (x_rot_active or y_rot_active or z_rot_active):
```

File: src/Editor/Scripts/Gizmo.gd [hunk 11]
```diff
@@ -5,96 +5,164 @@ var x_active := false
...
+			Vector3(0, 1, 0) if y_rot_active else \
+			Vector3(0, 0, 1)
+		
+		var plane = Plane(axis, get_parent().translation.dot(axis))
+		
+		var ray_old_mouse_position = camera.project_ray_normal(event.position - event.relative)
+		var ray_new_mouse_position = camera.project_ray_normal(event.position)
+		
+		var intersection_old = plane.intersects_ray(camera.translation, ray_old_mouse_position)
+		var intersection_new = plane.intersects_ray(camera.translation, ray_new_mouse_position)
```
Comment (HaSa1002) [inline src/Editor/Scripts/Gizmo.gd]:
```suggestion
		var camera := get_viewport().get_camera()
		
		var axis := \
			Vector3(1, 0, 0) if x_rot_active else \
			Vector3(0, 1, 0) if y_rot_active else \
			Vector3(0, 0, 1)
		
		var plane := Plane(axis, get_parent().translation.dot(axis))
		
		var ray_old_mouse_position := camera.project_ray_normal(event.position - event.relative)
		var ray_new_mouse_position := camera.project_ray_normal(event.position)
		
		var intersection_old := plane.intersects_ray(camera.translation, ray_old_mouse_position)
		var intersection_new := plane.intersects_ray(camera.translation, ray_new_mouse_position)
```

File: src/Editor/Scripts/Gizmo.gd [hunk 12]
```diff
@@ -5,96 +5,164 @@ var x_active := false
...
+		
+		# Do nothing when the mouse pointer doesn't hover over the plane
+		if intersection_old != null and intersection_new != null:
+			var diff = (intersection_old - get_parent().translation).signed_angle_to(intersection_new - get_parent().translation, axis)
```
Comment (HaSa1002) [inline src/Editor/Scripts/Gizmo.gd]:
```suggestion
			var diff := (intersection_old - get_parent().translation).signed_angle_to(intersection_new - get_parent().translation, axis)
```

File: src/Editor/Scripts/Gizmo.gd [hunk 13]
```diff
@@ -5,96 +5,164 @@ var x_active := false
...
+			get_parent().rotate(axis, diff)
+
+	elif event is InputEventMouseMotion and not any_axis_active():
```
Comment (HaSa1002) [inline src/Editor/Scripts/Gizmo.gd]:
```suggestion
	elif mm != null and not any_axis_active():
```

File: src/Editor/Scripts/Gizmo.gd [hunk 14]
```diff
@@ -5,96 +5,164 @@ var x_active := false
...
+				"y-axis":
+					$"y-axis/MeshInstance".get_surface_material(0).emission = Color(0.5, 1, 0.5)
+				"z-axis":
+					$"z-axis/MeshInstance".get_surface_material(0).emission = Color(0.5, 0.5, 1)
+				"x-rot":
+					$"x-rot/MeshInstance".get_surface_material(0).emission = Color(1, 0.5, 0.5)
+				"y-rot":
+					$"y-rot/MeshInstance".get_surface_material(0).emission = Color(0.5, 1, 0.5)
+				"z-rot":
+					$"z-rot/MeshInstance".get_surface_material(0).emission = Color(0.5, 0.5, 1)
```
Comment (HaSa1002) [inline src/Editor/Scripts/Gizmo.gd]: Please create export vars for these colours :>

File: src/Editor/Scripts/Gizmo.gd [hunk 15]
```diff
@@ -5,96 +5,164 @@ var x_active := false
...
+func _raycast_on_gizmo_layer() -> Dictionary:
+	var camera = get_viewport().get_camera()
+	
+	var ray_length = 1000
+	var mouse_pos = get_viewport().get_mouse_position()
+	var from = camera.project_ray_origin(mouse_pos)
+	var to = from + camera.project_ray_normal(mouse_pos) * ray_length
+	
+	var space_state = get_world().get_direct_space_state()
+	return space_state.intersect_ray(from, to, [  ], 0b10)
```
Comment (HaSa1002) [inline src/Editor/Scripts/Gizmo.gd]:
```suggestion
	var camera := get_viewport().get_camera()
	
	var ray_length := 1000
	var mouse_pos := get_viewport().get_mouse_position()
	var from := camera.project_ray_origin(mouse_pos)
	var to := from + camera.project_ray_normal(mouse_pos) * ray_length
	
	var space_state := get_world().get_direct_space_state()
	return space_state.intersect_ray(from, to, [ ], 0b10)
```

File: src/Editor/Scripts/Gizmo.gd [hunk 16]
```diff
@@ -5,96 +5,164 @@ var x_active := false
...
+
+
+func _reset_colors():
+	$"x-axis/MeshInstance".get_surface_material(0).emission = Color(1, 0, 0)
+	$"y-axis/MeshInstance".get_surface_material(0).emission = Color(0, 1, 0)
+	$"z-axis/MeshInstance".get_surface_material(0).emission = Color(0, 0, 1)
+	$"x-rot/MeshInstance".get_surface_material(0).emission = Color(1, 0, 0)
+	$"y-rot/MeshInstance".get_surface_material(0).emission = Color(0, 1, 0)
+	$"z-rot/MeshInstance".get_surface_material(0).emission = Color(0, 0, 1)
```
Comment (HaSa1002) [inline src/Editor/Scripts/Gizmo.gd]: Colours as export vars, please

**Review by HaSa1002 — CHANGES_REQUESTED** [review]
Gizmo is looking good, behaving as expected. Just some typing stuff and exporting the gizmo colours as variables. Your code is fine. We are dealing with 2d to 3d projection after all and it's not that long. Not at all, in fact. Nobody said the "easy" stuff is easy and short :>

**Comment by Jomosoto** [comment]
👍 Thanks for the review and sorry for my late reply. I'm not at home for the next two weeks, but I'll do the changes right when I'm back.

**Comment by HaSa1002** [comment]
> 👍 Thanks for the review and sorry for my late reply. I'm not at home for the next two weeks, but I'll do the changes right when I'm back.

Have great holidays

File: src/Editor/Modules/Gizmo.tscn (hunk 1 above)
Comment (Jomosoto) [reply src/Editor/Modules/Gizmo.tscn]: Yeah, I imported a bigger shape for the rotation circle hitbox, generated the collider and deleted the shape. Is this ok?

Comment (HaSa1002) [reply src/Editor/Modules/Gizmo.tscn]: You can try saving the shape to disk but, yeah that's fine

Comment (Jomosoto) [reply src/Editor/Modules/Gizmo.tscn]: I saved it to disk :+1:. I wasn't sure where to save it. I put it next the scene and script files.

**Review by HaSa1002 — APPROVED** [review]
lgtm

**Comment by HaSa1002** [comment]
Thanks ❤️

**Pull request merged**